Logic for improved guake like terminal functionality

maximstewart opened the pull request:
I wanted to add an improved Guake like terminal experience that adds options to lock the window to the top, bottom, left, or right part of any monitor. In addition, I wanted to override the toggle button so that the user could bind multiple terminals to separate toggle keys. This is on par with what Stjerm does which is abandoned at this time. 


Edit:  Fixed a stupid mistake after code cleanup. 

Edit: I found the request I made some time back:
https://github.com/gnome-terminator/terminator/issues/298

Yes, it was from an old account that no longer exists. Anyway, this is intended to address the above closed issue/feature request.

mattrose commented:
I really like this PR, and I'll look at it as soon as I can, but I'm kinda busy right now, so I'm not sure when I'll be able to look at it, but hopefully within the next week or so.

Thanks for contributing!

maximstewart commented:
I did find another weird bug where terminator itself couldn't seem to launch a new instance with the guake width and height arguments because it itself passed a string instead of int type. I changed it to default to string and then do a conversion to int when pulling the guake width and height. Anyway, I believe I've resolved all known issues and pushed a clean and rebased commit.

maximstewart commented:
Just wanted to update that everything has been stable for me using this "patch" or feature. I've not found any other bugs or issues hat needed my addressing that I haven't already done. Obviously this is just on my system but I don't believe it will be different elsewhere.

mattrose commented:
Sorry, this is still on my radar, I just haven't had time to look at it yet.  I hope tomorrow I'll have a little bit of free time after doing housework

maximstewart commented:
@mattrose 

Here is as start command:

terminator --guake-key="F8" --guake-side left --guake-width 700 --guake-height 800

To address your other points:

Editing a config file forces you to open an editor and make changes. Having it this way allows you to quickly spawn an instance as needed. Adding it to the UI would seem to pose a similar issue in my mind. In addition, the main way I use this is through my start script for my system. If I add a config entry then it would seem I'd start  an terminator instance just to start another terminator instance. Though, I could be misunderstanding your intent.

maximstewart commented:
@mattrose 

I'm not familiar with how terminator config is structured; but, wouldn't setting this in the config mean that each time you run terminator it spawns new instances of the same pinned terminals?  If so, I most certainly want to keep it as cmd args.

mattrose commented:
I don''t think so.  terminator only uses one process, and just spawns new windows upon repeated calls of terminator (unless you specifically disable dbus).  I will play with it some and get back to you.

maximstewart commented:
Sounds good, I'll await your findings. But, to ask and address some other points, by adding this to the config variables I would be limited to binding 4 windows, correct? Also, to be frank, I'm unsure of what help text file and man page files to edit. Still working out the project structure a bit.

mattrose commented:
Hmm, this doesn't seem to work on my F34 desktop with wayland.  Does this require X11 to work?

maximstewart commented:
@mattrose   Based on the import section at line 29 It is. But, Keybinder could be fixed to work with Wayland at this point. I didn't really look too deep.

maximstewart commented:
I hope you are doing well. I am just checking in to see if you have done further testing?

maximstewart commented:
So, sorry for taking so long to get back to this but I added what I think are the needed changes for the man page. I git rebased my commit and added changes to doc/terminator.1. I also did do some slight code cleanup/variable rename as well but nothing major. In addition, I updated some verbiage for terminatorlib/optionparse.py and added an import for RawTextHelpFormatter which allows the argparse add_argument method to format help text as expected when using \\n. (I can remove if not desired but otherwise the example text is long and on the same line as the description.) I found that fix the following SA post:

https://stackoverflow.com/questions/3853722/how-to-insert-newlines-on-argparse-help-text

maximstewart commented:
Ngl, I almost gave up on this given the misery I had trying to pull new changes, resolve merge conflict, then totally bungle it up, and thereafter try reverting my bungled merge only to revert my work but still pull in the newer changes. I got lucky that github shows the commits even if not tied to a branch.

maximstewart commented:
🥲

mattrose commented on terminatorlib/window.py:
```diff
@@ -92,15 +96,44 @@ def __init__(self):
+                    err('Window::__init__: Unable to parse geometry: %s' %
                             options.geometry)
 
+            if options.guake_key:
```
Can we move these command-line options to config file variables?  We can still leave them as command-line options, but merge them into the config file variables with self.config.options_get()

mattrose reviewed (changes requested):
The rest of this is ... fine, the whitespace changes are annoying, but I'd really have these be config variables as first-class, with command-line options being variable.

I won't worry about integrating this into the preferences dialog right now, but that's something that should be done down the road.

mattrose commented:
I'm sorry to be so picky.  I like your changes, and I think they bring out some functionality, but I really think having these as config variables is a much, much better solution for most people.
